File: firebase/functions/lib/events/live_meetings/get_user_id_from_agora_id.dart
import 'dart:async';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../../on_call_function.dart';
import '../../utils/firestore_utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/user/public_user_info.dart';

import '../../utils/utils.dart';

class GetUserIdFromAgoraId extends OnCallMethod<GetUserIdFromAgoraIdRequest> {
  GetUserIdFromAgoraId()
      : super(
          'GetUserIdFromAgoraId',
          (jsonMap) => GetUserIdFromAgoraIdRequest.fromJson(jsonMap),
        );

  @override
  Future<Map<String, dynamic>> action(
    GetUserIdFromAgoraIdRequest request,
    CallableContext context,
  ) async {
    orElseUnauthorized(context.authUid != null);

    final docs = await firestore
        .collection('publicUser')
        .where("agoraId", isEqualTo: request.agoraId)
        .get();

    final userDoc = docs.documents.firstOrNull;
    if (!(userDoc?.exists ?? false)) {
      throw HttpsError(
        HttpsError.notFound,
        'User with agora ID ${request.agoraId} not found',
        null,
      );
    }

    final userInfo = PublicUserInfo.fromJson(
      firestoreUtils.fromFirestoreJson(userDoc?.data.toMap() ?? {}),
    );

    return GetUserIdFromAgoraIdResponse(
      userId: userInfo.id,
    ).toJson();
  }
}

File: firebase/functions/src/events/live_meetings/get_user_id_from_agora_id.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../../on_call_function';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';
import { orElseUnauthorized, orElseNotFound } from '../../utils/utils';

interface GetUserIdFromAgoraIdRequest {
  agoraId: number | string;
}

interface GetUserIdFromAgoraIdResponse {
  userId?: string;
}

export class GetUserIdFromAgoraId extends OnCallMethod<GetUserIdFromAgoraIdRequest> {
  constructor() {
    super('GetUserIdFromAgoraId', (jsonMap) => jsonMap as GetUserIdFromAgoraIdRequest);
  }

  async action(
    request: GetUserIdFromAgoraIdRequest,
    context: functions.https.CallableContext
  ): Promise<GetUserIdFromAgoraIdResponse> {
    orElseUnauthorized(context.auth?.uid != null);

    const docs = await firestore
      .collection('publicUser')
      .where('agoraId', '==', request.agoraId)
      .get();

    if (docs.empty) {
      throw new functions.https.HttpsError('not-found', `User with agora ID ${request.agoraId} not found`);
    }

    const userInfo = firestoreUtils.fromFirestoreJson(docs.docs[0].data()) as unknown as { id?: string };

    return { userId: userInfo.id };
  }
}
